api/auth_middleware_test.go

```diff
@@ -1,63 +1,101 @@
+			expectedResponseHeaders: map[string]string{"x-service-account-name": "keyPairUser"},
+			expectedResponseCode:    http.StatusOK,
+		},
+		{
+			testName:                "OAuthAuthorization",
+			serviceAccount:          oauthSA,
+			requestHeaders:          map[string]string{"authorization": fmt.Sprintf("bearer %s", token.AccessToken)},
+			expectedResponseHeaders: map[string]string{"x-email": oauthSA.Email},
+			expectedResponseCode:    http.StatusOK,
+		},
+		{
+			testName:                "WrongOAuthTokenAuthorization",
+			serviceAccount:          oauthSA,
+			requestHeaders:          map[string]string{"authorization": fmt.Sprintf("bearer %s", "wrong token")},
+			expectedResponseHeaders: emptyMap,
+			expectedResponseCode:    http.StatusUnauthorized,
+		},
+		{
+			testName:                "WrongKeyPairAuthorization",
+			serviceAccount:          keyPairSA,
+			requestHeaders:          map[string]string{"authorization": fmt.Sprintf("keypair %s:%s", "wrong key_id", "wrong_key_secret")},
+			expectedResponseHeaders: emptyMap,
+			expectedResponseCode:    http.StatusUnauthorized,
+		},
+		{
+			testName:                "UndefinedAuthorization",
+			serviceAccount:          nil,
+			requestHeaders:          emptyMap,
+			expectedResponseHeaders: emptyMap,
+			expectedResponseCode:    http.StatusUnauthorized,
+		},
+		{
+			testName:                "UnsupportedAuthorization",
+			serviceAccount:          oauthSA,
+			requestHeaders:          map[string]string{"authorization": fmt.Sprintf("basic %s", b64.StdEncoding.EncodeToString([]byte("user:password")))},
+			expectedResponseHeaders: emptyMap,
+			expectedResponseCode:    http.StatusUnauthorized,
+		},
 	}
-}
 
-func TestAuthMiddlewareBearerShouldAuthenticateUser(t *testing.T) {
-	helpers.CleanupPG(t)
+	for _, testCase := range testCases {
+		t.Run(testCase.testName, func(t *testing.T) {
+			app := helpers.GetApp(t)
+			req, err := http.NewRequest(http.MethodGet, "/service_accounts", nil)
 
-	rootSA := helpers.CreateRootServiceAccountWithOAuth(t)
-	tokensRepo := helpers.GetRepo(t).Tokens
-	tokens, _ := tokensRepo.FindByEmail(rootSA.Email)
-	token := tokens[0]
-	app := helpers.GetApp(t)
-	req, err := http.NewRequest(http.MethodGet, "/service_accounts", nil)
+			if err != nil {
+				t.Errorf("Could not create HTTP request")
+			}
 
-	if err != nil {
-		t.Errorf("Could not create HTTP request for /service_accounts")
-	}
+			for header, content := range testCase.requestHeaders {
+				req.Header.Set(header, content)
+			}
 
-	req.Header.Set("Authorization", fmt.Sprintf("bearer %s", token))
-	response := helpers.DoRequest(t, req, app.GetRouter())
+			response := helpers.DoRequest(t, req, app.GetRouter())
 
-	if response.Code != http.StatusOK {
-		t.Errorf("Expected status %d. Got %d", http.StatusOK, response.Code)
-	}
+			if response.Code != testCase.expectedResponseCode {
+				t.Errorf("Expected status %d. Got %d", testCase.expectedResponseCode, response.Code)
```

Comment:
Use t.Fatalf here? I expect everything else will fail if we got the wrong error code.

Reply:
It mostly sure will, but it may be good to also see the response's header in case of failure as it may help identify the error.

Reply:
Sure, tweak the message as you see fit.